pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/loader/defaultcolumn/BaseDefaultColumnHandler.java

```diff
@@ -571,21 +573,53 @@ private void createDerivedColumnV1Indices(String column, FunctionEvaluator funct
       int numDocs = _segmentMetadata.getTotalDocs();
       Object[] outputValues = new Object[numDocs];
       PinotDataType outputValueType = null;
+
+      FieldSpec fieldSpec = _schema.getFieldSpecFor(column);
+      Object defaultNullValue = fieldSpec.getDefaultNullValue();
+      // Just log the first function evaluation error
+      int functionEvaluateErrorCount = 0;
+      Exception functionEvalError = null;
+      Object[] inputValuesWithError = null;
+
       for (int i = 0; i < numDocs; i++) {
         for (int j = 0; j < numArguments; j++) {
           inputValues[j] = valueReaders.get(j).getValue(i);
         }
-        Object outputValue = functionEvaluator.evaluate(inputValues);
-        outputValues[i] = outputValue;
-        if (outputValueType == null) {
-          Class<?> outputValueClass = outputValue.getClass();
-          outputValueType = FunctionUtils.getArgumentType(outputValueClass);
-          Preconditions.checkState(outputValueType != null, "Unsupported output value class: %s", outputValueClass);
+        try {
+          Object outputValue = functionEvaluator.evaluate(inputValues);
+          outputValues[i] = outputValue;
+          if (outputValueType == null) {
+            Class<?> outputValueClass = outputValue.getClass();
+            outputValueType = FunctionUtils.getArgumentType(outputValueClass);
+            Preconditions.checkState(outputValueType != null, "Unsupported output value class: %s", outputValueClass);
+          }
+        } catch (Exception e) {
+          if (!errorOnFailure) {
+            functionEvaluateErrorCount++;
+            // Just record one error to log after the loop
+            if (functionEvalError == null) {
+              functionEvalError = e;
+              inputValuesWithError = Arrays.copyOf(inputValues, inputValues.length);
+            }
+            outputValues[i] = defaultNullValue;
```

Comment:
We cannot directly put default value here. Say scalar function returns `TIMESTAMP`, the field is configured as `LONG`, when converting the type, it will expect `Timestamp` value instead of `Long` value.
The proper solution is to skip filling the `outputValue` here and do `null` check afterwards.